### PR TITLE
Feature/even more code smells

### PR DESCRIPTION
**Pull request by mgovers**
another removal of quite a number of (critical and major) code smells

* `using raw_data_ptr = void*` [cpp:S5008](https://sonarcloud.io/organizations/powergridmodel/rules?open=cpp%3AS5008&rule_key=cpp%3AS5008) (e.g.: https://sonarcloud.io/project/issues?open=AYcsrsmqzfpi9utDw90O&id=PowerGridModel_power-grid-model)
* capture explicitly instead of `[&]` [cpp:S3608](https://sonarcloud.io/organizations/powergridmodel/rules?open=cpp%3AS3608&rule_key=cpp%3AS3608) (e.g.: https://sonarcloud.io/project/issues?open=AYcsrsoTzfpi9utDw92D&id=PowerGridModel_power-grid-model)
* use `explicit` [cpp:S1709](https://sonarcloud.io/organizations/powergridmodel/rules?open=cpp%3AS1709&rule_key=cpp%3AS1709) (https://sonarcloud.io/project/issues?open=AYcsrsnVzfpi9utDw91a&id=PowerGridModel_power-grid-model)
* fix clang-tidy warning `modernize-deprecated-headers`

### CODE REVIEW

**Comment by mgovers** [comment]
new code smells are either other ones that were already reported or i moved the majority of cases of n issue type and contained them in one specific location

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/meta_data.hpp [hunk 1]
```diff
@@ -26,13 +26,17 @@ struct trait_pointer_to_member<ValueType StructType::*> {
     using value_type = ValueType;
 };
 
-using SetNaNFunc = std::add_pointer_t<void(void*)>;
-using CheckNaNFunc = std::add_pointer_t<bool(void const*)>;
-using SetValueFunc = std::add_pointer_t<void(void*, void const*)>;
-using CompareValueFunc = std::add_pointer_t<bool(void const* ptr_x, void const* ptr_y, double atol, double rtol)>;
+using raw_data_ptr = void*;              // raw mutable data ptr
+using raw_data_const_ptr = void const*;  // raw read-only data ptr
```
Comment (TonyXiang8787) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/meta_data.hpp:30]: I would use CamelCase for these two pointer types.

File: power_grid_model_c/power_grid_model_c/include/power_grid_model_c.h [hunk 2]
```diff
@@ -39,8 +39,13 @@
 // API_MACRO_BLOCK
 
 // integers
+#ifdef __cplusplus
+#include <cstddef>
+#include <cstdint>
+#else
 #include <stddef.h>
 #include <stdint.h>
+#endif
```
Comment (TonyXiang8787) [inline power_grid_model_c/power_grid_model_c/include/power_grid_model_c.h:48]: Do we really need this? What is the down side to include C header in C++?

Comment (mgovers) [reply power_grid_model_c/power_grid_model_c/include/power_grid_model_c.h:48]: the downside is that everything that is `std` in the `<cstdxxx>` libraries becomes global namespace in `<stdxxx.h>`. that means that users including our C API using C++ code will have their environment polluted with global namespace features. We should aim to keep our users' environment as clean as possible.

**Review by TonyXiang8787 — APPROVED** [review]

**Pull request merged**